Android/app/src/androidTest/java/tie/hackathon/travelguide/Tests/Helpers.java

```diff
@@ -0,0 +1,261 @@
+                final Matcher<View> viewMatcher = withId(viewId);
+
+                do {
+                    for(View child : TreeIterables.breadthFirstViewTraversal(view)) {
+                        // found view with required ID
+                        if(viewMatcher.matches(child)) {
+                            return;
+                        }
+                    }
+                    uiController.loopMainThreadForAtLeast(50);
+                }
+                while(System.currentTimeMillis() < endTime);
+
+                // timeout happens
+                throw new PerformException.Builder()
+                        .withActionDescription(this.getDescription())
+                        .withViewDescription(HumanReadables.describe(view))
+                        .withCause(new TimeoutException())
+                        .build();
+            }
+        };
+    }
+
+    public static void waitFor(int rid, int time) throws Exception {
+        onView(isRoot()).perform(waitId(rid, TimeUnit.SECONDS.toMillis(time)));
+    }
+
+    public static void clickASpecificObjectWithTextAndId(String text, int rid) throws Exception {
+        onView(allOf(withText(text), withId(rid))).perform(click());
+    }
+
+    public static void swipeLeftASpecificObjectWithTextAndId(String text, int rid) throws Exception {
+        onView(allOf(withText(text), withId(rid))).perform(swipeLeft());
+    }
+
+    public static void swipeRightASpecificObjectWithTextAndId(String text, int rid) throws Exception {
+        onView(allOf(withText(text), withId(rid))).perform(swipeRight());
+    }
+
+    public static void clickASpecificObjectWithId(int rid) throws Exception {
+        onView(withId(rid)).perform(click());
+    }
+
+    //Matches the first object found when multiple objects with the same id/text/index are available
+    public static Matcher<View> first(final Matcher<View> expected) {
+
+        return new TypeSafeMatcher<View>() {
+            private boolean first = false;
+
+            @Override
+            protected boolean matchesSafely(View item) {
+
+                if(expected.matches(item) && !first) {
+                    return first = true;
+                }
+
+                return false;
+            }
+
+            @Override
+            public void describeTo(Description description) {
+                description.appendText("Matcher.first( " + expected.toString() + " )");
+            }
+        };
+    }
+
+    public static void clickOnTheFirstObjectInTheListWhenAmbiguous(String text, int rid) throws Exception {
+        //onView( withId( R.id.interest_2 ) ).perform( RecyclerViewActions.scrollTo( first(withText("View on Map"))));
+        onView(Helpers.first(allOf(withText(text), withId(rid)))).perform(click());
+    }
+
+    public static void navigateBackToApp() throws Exception {
+        UiDevice mDevice = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation());
```

Comment:
Nu avem deja "device" in "EspressoTestBase"?

Reply:
ba da :(, scot afara.